File: Python2/n-ary_tree_postorder_traversal.py
class Node(object):
    def __init__(self, val, children):
        self.val = val
        self.children = children


class Solution(object):
    def postorder(self, root):
        if not root:
            return []
        output = []
        stack = [root]
        while stack:
            this_node = stack.pop()
            output.append(this_node.val)
            if this_node.children:
                stack.extend(this_node.children)
        return output[::-1]


if __name__ == '__main__':
    tree = \
        Node(1,
             [Node(3, [Node(5, []), Node(6, [])]), Node(2, []), Node(4, [])])
    sol = Solution()
    print sol.preorder(tree)

File: C++/n-ary_tree_postorder_traversal.cpp
// This problem taught me: no silly solution
// Pursuing the best solution blindly is not practical

#include "./n-ary_tree_postorder_traversal.h"
#include <stack>
#include <iostream>

using namespace std;

vector<int> Solution::postorder(Node* root) {
    vector<int> output;
    if (root == NULL) {
        return output;
    }
    stack<Node*> temp;
    vector<Node*>::const_iterator iter;
    temp.push(root);
    // Depth-Fisrt Search postorder
    while (!temp.empty()) {
        Node* this_node = temp.top();
        output.push_back(this_node->val);
        temp.pop();
        for (iter = this_node->children.cbegin(); iter != this_node->children.cend(); ++iter) {
            temp.push((*iter));
        }
    }
    // Just reverse it!
    // Linear in half the distance between first and last: Swaps elements.
    // O(n/2)
    reverse(output.begin(), output.end());
    return output;
}


int main() {
    Node tree;
    tree.val = 1;
    tree.children.push_back(new Node());
    tree.children.back()->val = 3;
    tree.children.back()->children.push_back(new Node());
    tree.children.back()->children.back()->val = 5;
    tree.children.back()->children.push_back(new Node());
    tree.children.back()->children.back()->val = 6;
    tree.children.push_back(new Node());
    tree.children.back()->val = 2;
    tree.children.push_back(new Node());
    tree.children.back()->val = 4;

    Solution* sol = new Solution();
    vector<int> output = sol->postorder(&tree);
    //Utility::printVector(output);
    return 0;
}